### PR TITLE
Fix `LightningDataModule` hparams parsing

### LINKED ISSUE #12768: tuner.scale_batch_size does not update DataModule hparams
## 🐛 Bug

`tuner.scale_batch_size` finds the suitable batch size and update the batch size of the model AND datamodule.
For the model, `tuner.scale_batch_size` updates the batch size in the model regardless of `model.batch_size` and `model.hparams.batch_size`.

However, for the datamodule, `tuner.scale_batch_size` updates `datamodule.batch_size` only, and keep `datamodule.hparams.batch_size` unchanged if hparams are in use.

So the tuner behaves differently for model and datamodule, which could be a trigger of confusion for users.

### To Reproduce

https://gist.github.com/tsuga/289febb6cfe8c05e6c41cce043147174

### Expected behavior

The tuner should updates `datamodule.hparams.batch_size`

### Environment
* CUDA:
        - GPU:
                - Tesla P100-PCIE-16GB
        - available:         True
        - version:           11.1
* Packages:
        - numpy:             1.21.5
        - pyTorch_debug:     False
        - pyTorch_version:   1.10.0+cu111
        - pytorch-lightning: 1.6.1
        - tqdm:              4.64.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - 
        - processor:         x86_64
        - python:            3.7.13
        - version:           #1 SMP Tue Dec 7 09:58:10 PST 2021

### Additional context

cc @akihironitta @borda @rohitgr7 @carmocca @awaelchli @ananthsub @ninginthecloud @jjenniferdai

### PR DESCRIPTION
**Pull request by tanmoyio**
## What does this PR do?

Updates DataModule hparams for `tuner.scale_batch_size()` 

Fixes #12768

The actual fix was that we need to correctly parse datamodule hparams too. check the review for a better understanding.

### Does your PR introduce any breaking changes? If yes, please list them.

No

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by tanmoyio** [comment]
@awaelchli I have tried this, without touching the `_lightning_get_all_attr_holders`, and this seems to work.

**Comment by awaelchli** [comment]
This looks great! Would you mind extending the test test_auto_scale_batch_size_set_model_attribute in `tests/tuner/test_scale_batch_size.py` with a call to `self.save_hyperparameters` in the datamodule and adding the necessary assertion?

**Comment by tanmoyio** [comment]
@awaelchli sure

**Comment by akihironitta** [comment]
Hi @tanmoyio! Thanks for working on this!

Here's the TODO list in this PR!
- [ ] Add a changelog entry
- [ ] Extend the existing test to cover the change of this PR (https://github.com/PyTorchLightning/pytorch-lightning/pull/12806#issuecomment-1107933705)

Let us know if you need help!

**Comment by tanmoyio** [comment]
@[akihironitta](https://github.com/akihironitta) I have added the unit tests, let me know if I need to modify anything, then I modify the changelog when everything is ready to go.

**Comment by akihironitta** [comment]
@tanmoyio Would you mind having a look at the test failures? Also, could you update the changelog?

**Comment by tanmoyio** [comment]
@akihironitta sorry, I should have done that earlier, will take a look into that.

**Comment by tanmoyio** [comment]
@akihironitta is it fine now ?

**Comment by rohitgr7** [comment]
I think this PR needs some more work.
- What if batch_arg_name is different in model and datamodule?
- What if the argname is not present in both datamodule and model?
- Maybe a few more edge cases...

do you want to work on it or I can push updates? @tanmoyio

**Comment by tanmoyio** [comment]
@rohitgr7 
Alright, these are the probable scenarios I can think of, for hparams
1. `batch_size` is present in both
2. `batch_size` is present in one of datamodule/model
3. `batch_size` is present in none of datamodule/model
4.  what if `batch_arg_name` is different in both model/datamodule 
5.  what if `batch_arg_name` is same in both model/datamodule but not `batch_size`

I will work on the PR, let me know if there is anything else I need to include

**Comment by rohitgr7** [comment]
@tanmoyio 

we should not consider `batch_size` anywhere and only rely on `batch_arg_name` for all the cases.

File: src/pytorch_lightning/utilities/parsing.py [hunk 1]
```diff
@@ -321,14 +321,17 @@ def _lightning_get_all_attr_holders(model: "pl.LightningModule", attribute: str)
...
     trainer = model._trainer
     # Check if the attribute in datamodule (datamodule gets registered in Trainer)
-    if trainer is not None and trainer.datamodule is not None and hasattr(trainer.datamodule, attribute):
-        holders.append(trainer.datamodule)
+    if trainer is not None and trainer.datamodule is not None:
+        if hasattr(trainer.datamodule, attribute):
+            holders.append(trainer.datamodule)
+
+        if hasattr(trainer.datamodule, "hparams") and attribute in trainer.datamodule.hparams:
+            holders.append(trainer.datamodule.hparams)
```
Comment (rohitgr7) [inline src/pytorch_lightning/utilities/parsing.py:334]: this is the fix, we consider datamodule hparams

File: tests/tests_pytorch/tuner/test_scale_batch_size.py [hunk 2]
```diff
@@ -140,47 +140,64 @@ def test_auto_scale_batch_size_trainer_arg(tmpdir, scale_arg):
     assert not os.path.exists(tmpdir / "scale_batch_size_temp_model.ckpt")
 
 
-@RunIf(min_cuda_gpus=1)
```
Comment (rohitgr7) [inline tests/tests_pytorch/tuner/test_scale_batch_size.py:143]: no need for run on GPUs

File: tests/tests_pytorch/tuner/test_scale_batch_size.py [hunk 3]
```diff
@@ -140,47 +140,64 @@ def test_auto_scale_batch_size_trainer_arg(tmpdir, scale_arg):
...
+    before_batch_size = hparams["batch_size"]
 
-    class HparamsBatchSizeModel(BatchSizeModel):
+    class HparamsBatchSizeModel(BoringModel):
         def __init__(self, *args, **kwargs):
-            super().__init__(*args, **kwargs)
+            super().__init__()
             self.save_hyperparameters()
 
-        def dataloader(self, *args, **kwargs):
```
Comment (rohitgr7) [inline tests/tests_pytorch/tuner/test_scale_batch_size.py:157]: this was redundant, we now initialize the dataloaders explictly.

File: tests/tests_pytorch/tuner/test_scale_batch_size.py [hunk 4]
```diff
@@ -140,47 +140,64 @@ def test_auto_scale_batch_size_trainer_arg(tmpdir, scale_arg):
...
+
+    trainer = Trainer(default_root_dir=tmpdir, max_epochs=1, auto_scale_batch_size=True)
+    trainer.tune(model, scale_batch_size_kwargs={"steps_per_trial": 2, "max_trials": 4})
+    after_batch_size = model.hparams.batch_size if use_hparams else model.batch_size
+    assert before_batch_size != after_batch_size
+    assert after_batch_size <= len(trainer.train_dataloader.dataset)
+
+
+@pytest.mark.parametrize("use_hparams", [True, False])
+def test_auto_scale_batch_size_set_datamodule_attribute(tmpdir, use_hparams):
```
Comment (rohitgr7) [inline tests/tests_pytorch/tuner/test_scale_batch_size.py:173]: this one is for datamodule.

File: tests/tests_pytorch/utilities/test_parsing.py [hunk 5]
```diff
@@ -116,10 +126,11 @@ def test_lightning_getattr():
         value = lightning_getattr(m, "learning_rate")
         assert value == i, "attribute not correctly extracted"
 
-    model5, model6, model7 = models[4:]
+    model5, model6, model7, model8 = models[4:]
     assert lightning_getattr(model5, "batch_size") == 8, "batch_size not correctly extracted"
     assert lightning_getattr(model6, "batch_size") == 8, "batch_size not correctly extracted"
     assert lightning_getattr(model7, "batch_size") == 8, "batch_size not correctly extracted"
+    assert lightning_getattr(model8, "batch_size") == 2, "batch_size not correctly extracted"
```
Comment (rohitgr7) [inline tests/tests_pytorch/utilities/test_parsing.py:133]: `2` because it uses a standalone trainer instance which doesn't update the `batch_size` argument of datamodule.

**Review by rohitgr7 — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

**Review by awaelchli — APPROVED** [review]
